Doc db migration guide

i-chvets opened the pull request:
# Description

For upgrade from 1.7 to 1.7 Patch 1 DB migration instructions need to be included.This can be provided in form of Discourse post and linked in upgrade guide or as a part of upgrade guide. Instructions should include doing back up, unrelate, import back up into mysql, relate.

The instructions in this document can be tested as shell commands and can be automatically tested, if required. A separate work item is required for this.

NOTE
CKF 1.8 does NOT work with MariaDB. This means clients need to:
- First migrate the DB, to MySQL
- Then do the CKF 1.7 to 1.8 upgrade

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "source": [
+    "# Charmed Kubeflow Databases Migration Guide\n",
+    "\n",
+    "There are two databases in Charmed Kubeflow - `katib-db` and `kfp-db`. In Charmed Kubeflow 1.7 these MariaDB databases deployed by `charmed-osm-mariadb-k8s` charm. Since release 1.7 there is an option to relate charm to new MySQL database that is deployed by `mysql-k8s` charm. In cases where previous deployment has MariaDB a migration of database is required."
```
I'd do something like this:

```suggestion
    "There are two databases in Charmed Kubeflow - `katib-db` and `kfp-db` which are backed by the MariaDB `charmed-osm-mariadb-k8s` charm. In Charmed Kubeflow 1.7 we introduced support for MySQL databases, which can be deployed using the `mysql-k8s` charm.
This document will guide you through the migration process from a MariaDB to a MySQL database."
```

ColmBhandal replied:
Let's not reference "This document". Here's my suggestion:

When upgrading to Charmed Kubeflow 1.7, Maria DB databases must be migrated to MySQL. There are two databases in Charmed Kubeflow - `katib-db` and `kfp-db`. Prior to 1.7, both were MariaDB databases backed by the `charmed-osm-mariadb-k8s` charm. In 1.7 they are backed by MySQL the `mysql-k8s` charm.

i-chvets replied:
I will modif.

i-chvets replied:
Updated.

DnPlas replied:
I think this still applies

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "source": [
+    "## Prerequisites\n",
+    "\n",
+    "In order to perform migaraion the following tools should be installed on client machine that has access to Charmed Kubeflow deployment:\n",
```
```suggestion
    "In order to perform migration the following tools should be installed on client machine that has access to Charmed Kubeflow deployment:\n",
```

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "juju remove-relation katib-db-manager katib-db\n",
+    "\n",
+    "# connect to unit's IP address using mysql client and verify that all data is intact\n",
+    "mysql --host=$KATIB_DB_IP --user=$KATIB_DB_USER --password=$KATIB_DB_PASSWORD\n",
+    "\n",
+    "# create backup of all databases file using `mysqldump` utility, username, password, and unit's IP address\n",
+    "# onbtained earlier\n",
+    "mysqldump --host=$KATIB_DB_IP --user=$KATIB_DB_USER --password=$KATIB_DB_PASSWORD --column-statistics=0 --databases katib  > katib-db.sql\n",
```
These commands render weirdly:

![image](https://github.com/canonical/bundle-kubeflow/assets/18201737/37116c91-971b-47c6-b9e0-921fd1e7d8d3)

ColmBhandal replied:
That is bizarre. I'm seeing the same in GitHub preview. Same below.

i-chvets replied:
Yes, it is very strange in GitHub.

DnPlas replied:
I expect this to not be an issue when we publish this doc in charmed-kubeflow.io, right?

i-chvets replied:
Yes. It should be all good. VScode displays it properly.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "MYSQL_DB_IP=$(juju show-unit katib-db/0 | yq '.[] | .address')\n",
+    "\n",
+    "# connect to new DB using username, password, and unit's IP address and restore database from backup\n",
+    "mysql --host=$MYSQL_DB_IP --user=$MYSQL_DB_USER --password=$MYSQL_DB_PASSWORD < katib-db.sql\n",
```
This renders weirdly

![image](https://github.com/canonical/bundle-kubeflow/assets/18201737/d83df62c-e095-40ba-927f-94f26bee0159)

i-chvets replied:
Yes, it is very strange in GitHub.

DnPlas replied:
About the weird rendering, @ColmBhandal discovered that the text between $ is causing the text to be rendered as math context, related: https://github.com/jupyter/notebook/issues/1080

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Katib DB migration"
```
It looks like the Katib DB migration is exactly the same as the KFP DB migration, I think we could avoid duplication by just adding a generic set of steps and tell people that they work for both, just the name of the application and unit is going to be different.

ColmBhandal replied:
I'm not sure.

This actually raises a difficulty that we never thought about before in terms of testable docs. If our doc testing process behind the scenes simply runs all the code blocks against a CKF deployment, then it will force us to write docs in this way - with exhaustive concrete sequences of commands that the user must run. But this conflicts with the readability of the docs, which would read better as Daniela suggested.

So we either have to:

1. Sacrifice readability for the sake of testability
2. Sacrifice testability for the sake of readability
3. Implement a more complex doc testing process e.g. we could present abstract steps to the user

i-chvets replied:
Not really identical. We get credentials differently using different unit names and different commands. Also, I could not get Katib DB populated properly though, so there could be some additional work to be done when we have populated Katib DB, like porting constraints, etc.


RE: testability.
We can have sections of doc executed multiple times with different parameters. Eg. we can put something in comments to indicate to test script that different DB name can be used, for example. This is whole new Epic though.

ColmBhandal replied:
Cool. My vote is leave it as it is.

i-chvets replied:
@DnPlas I pushed the change. I think, it made it more confusing when we try to create generilize. Please review and let me know if this works. I am not convinced that this is better than outlining each DB migration for each DB.

DnPlas replied:
Alright, it's okay with me

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prerequisites\n",
```
I'd add a subsection before or after "Prerequisites" that helps users know under which circumstances they are supposed to migrate, for instance, I'd give instructions on how to check which database they have deployed and based on that they can check whether they need to migrate.

i-chvets replied:
I will use Colm's suggestions and just indicate this in the begging of the doc. Basically saying that when upgrading to 1.8 you must migrate databases if they are MariaDB. Will add the the check for type of DB in Prerequisites.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "# remove relation between kfp-api charm and MariaDB charm\n",
+    "juju remove-relation kfp-api kfp-db\n",
+    "\n",
+    "# connect to unit's IP address using mysql client and verify that all data is intact\n",
```
How would users know data is intact?

i-chvets replied:
Will add.

i-chvets replied:
Updated.

DnPlas replied:
I think this still applies

i-chvets replied:
Should provide SQL queries? How deep should we go into describing DB operations?
I'd would leave it outside of this guide, but open to suggestions.

DnPlas replied:
I commented because you had it initially, if you don't think it is required we are good

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "# deploy new MySQL database charm\n",
+    "juju deploy mysql-k8s kfp-db-mysql --channel 8.0/stable --trust\n",
+    "\n",
+    "# obtain username and password of new MySQL database from DB relation\n",
+    "# note that since relation is not yet established it is impossible to retrieve this information from that relation\n",
+    "# the relation can only be established after data is retored into new database\n",
```
This set of comments is a bit confusing, on one hand we are telling users to retrieve username and password from the relation data, but on the other hand we are telling that it is impossible to retrieve this information because the relation has not been established.

i-chvets replied:
Absolutely! That needs to be removed, because we are actually getting data from the charm. Will update.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "\n",
+    "# relate kfp-api and new MySQL database charm\n",
+    "juju relate kfp-api:relational-db kfp-db-mysql:database"
+   ]
```
Should we remove the *-db application from the model after we are done with the migration? For instance`juju remove-application kfp-db`

i-chvets replied:
Good point. Will add.

DnPlas reviewed (commented):
Thanks @i-chvets left some comments. I would like to run the commands on a live deployment to be able to approve the instructions.

I wonder what is the advantage of using a notebook instead of a regular markdown, since we are only executing shell commands.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+    "# 'root_password'\n",
+    "KATIB_DB_MYSQL_RELATION_ID=$(juju show-unit katib-db/0 | yq '.[] | .relation-info | select(.[].endpoint == \"mysql\") | .[0] | .relation-id')\n",
+    "KATIB_DB_USER=root\n",
+    "KATIB_DB_PASSWORD=$(bash -c \"juju run --unit katib-db/0 'relation-get -r $KATIB_DB_MYSQL_RELATION_ID - katib-db/0' | grep root_password\" | awk '{print $2}')\n",
```
I had an issue with this command:

```
KATIB_DB_PASSWORD=$(bash -c "juju run --unit katib-db/0 'relation-get -r $KATIB_DB_MYSQL_RELATION_ID - katib-db/0' | grep root_password" | awk '{print $2}')
ERROR option provided but not defined: --unit
```

DnPlas replied:
I removed `--unit`, but now I get

```
KATIB_DB_PASSWORD=$(bash -c "juju run katib-db/0 'relation-get -r $KATIB_DB_MYSQL_RELATION_ID - katib-db/0' | grep root_password" | awk '{print $2}')
ERROR invalid unit or action name "relation-get -r 14 - katib-db/0"
```

I am using juju 3.1/stable, though

i-chvets replied:
Strange. Need to verify.
But guide target migration from MariaDB, which won't run on juju 3.1. Do we want to specify what version of Juju should be used?
Also, do you think it is worth mentioning that first we do DB migration then we do Juju migration? Or should we leave it out of scope for this?

i-chvets replied:
Added Juju 2.9 as requirement. Command should work.

DnPlas replied:
I think that, since we are going to start supporting 3.1 soon, we should mention how to deal with both cases. Also, 1.8 only supports 3.1, which will require some instructions for what to do first. Should I upgrade my cluster first, then juju then do the DB migration?

i-chvets replied:
We have it in Jira, I will add it here: DB migration must be done prior to any other upgrade and it is all done useing Juju 2.9.

This makes mentioning of 3.1 out of scope for this guide. This should be left for overlall CKF upgrade guide.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prerequisites\n",
```
We may need to specify that the migration instructions could be slightly different if using juju 2.9 and 3.1

i-chvets replied:
I suspect we will be going with Juju 2.9, since that's what we used to deploy 1.7. MariaDB only exists in <=1.7.

DnPlas replied:
So the idea will be to first migrate then upgrade juju, in the case users require it?

i-chvets replied:
This should be mentioned in overall CKF guide (non existent at this point): DB migration shoud be done prior to CKF upgrade.

DnPlas replied:
I think it's worth mentioning with a big note saying that this step should be performed before even upgrading.

i-chvets replied:
Yes. And also need the same in upgrade guide.

ColmBhandal commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,137 @@
+   "source": [
+    "## Prerequisites\n",
+    "\n",
+    "In order to perform migaraion the following tools should be installed on client machine that has access to Charmed Kubeflow deployment:\n",
```
```suggestion
    "`mysql-client` must be installed on the machine that has access to Charmed Kubeflow. To install:",
```

Then add the code to install

i-chvets replied:
Updated.

DnPlas replied:
I believe this still applies.

i-chvets replied:
Updated.

ColmBhandal reviewed (commented):
Overall feedback - it reads more like a script than a doc. Since it's a Jupyter notebook, can we put the comments in as cells and split up the code into multiple cells?

i-chvets commented:
> I wonder what is the advantage of using a notebook instead of a regular markdown, since we are only executing shell commands.

If we are using notebook, we can load it as JSON into Python script and run test on it. Basically, we have a testable document. [Example of test](https://github.com/canonical/kubeflow-examples/blob/main/kubeflow-cos-integration/tests/test_kf_cos_doc.py)
The idea is to covert this notebook to markdown (which is very easy to do) after we approve this PR and it will serve as a nice document  in our repo. [Example of markdown](https://github.com/canonical/kubeflow-examples/blob/main/kubeflow-cos-integration/README.md) generated from [corresponding notebook](https://github.com/canonical/kubeflow-examples/blob/main/kubeflow-cos-integration/kubeflow-cos-integration.ipynb)

DnPlas commented:
> Overall feedback - it reads more like a script than a doc. Since it's a Jupyter notebook, can we put the comments in as cells and split up the code into multiple cells?

WeComments could be replaced with small paragraphs for giving context, and/or setting the expectation of the command to be run, then each command or related command could be one cell.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+   "outputs": [],
+   "source": [
+    "    # replace DB_CHARM with desired unit name: `katib-db` or `kfp-db`\n",
+    "    DB_CHARM= kfp-db | katib-db\n",
```
```suggestion
    "    DB_CHARM= <kfp-db | katib-db>\n",
```

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "\n",
+    "- Client machine with access to deployed Charmed Kubeflow 1.7.\n",
+    "- Juju version is 2.9.\n",
+    "- Verify that migration of databases is required by checking types of databases deployed. If database charm deployed is `mariadb-k8s` migration is required:"
```
I'm not sure about the wording, maybe something like:

```suggestion
    "- A database migration is required if the output of the following command is `mariadb-k8s` <something something>:"
```

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+   "metadata": {},
+   "source": [
+    "\n",
+    "- Enough storage in the cluster to support backup/restore of the databases.\n",
+    "- Enough storage to store database data for backup on client machine.\n",
```
What is the difference between these two?

i-chvets replied:
Yes, we can just leave note about storage on the cluser.

i-chvets replied:
Updated.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+   "outputs": [],
+   "source": [
+    "# replace DB_UNIT with desired unit name: `katib-db` or `kfp-db`\n",
+    "DB_UNIT= katib-db | kfp-db\n",
```
```suggestion
    "    DB_CHARM= <kfp-db | katib-db>\n",
```

i-chvets replied:
Will add.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "DB_UNIT= katib-db | kfp-db\n",
+    "\n",
+    "# obtain username and password of existing MariadDB database from DB relation\n",
+    "# in Charmed Kubeflow 1.6 and 1.7 username used is `root` and password is specified in `mysql` relation by\n",
```
I wouldn't mention CKF 1.6 as we'll stop supporting that version pretty soon.

i-chvets replied:
+1

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "# obtain username and password of existing MariadDB database from DB relation\n",
+    "# in Charmed Kubeflow 1.6 and 1.7 username used is `root` and password is specified in `mysql` relation by\n",
+    "# 'root_password'\n",
+    "DB_MYSQL_RELATION_ID=$(juju show-unit $DB_UNIT/0 | yq '.[] | .relation-info | select(.[].endpoint == \"mysql\") | .[0] | .relation-id')\n",
```
I think we should definitely have DB_UNIT to point to the unit (with number) as it can be any number, not just `app-name/0`

i-chvets replied:
Will update.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+   "source": [
+    "# ensure that there are no new connections are made and that dababase is not altered\n",
+    "# remove relation between katib-db-manager charm and katib-db charm\n",
+    "juju remove-relation katib-db-manager katib-db\n",
```
Should we remove the relation after we have performed the database backup?

i-chvets replied:
The goal of relation removal is to prevent DB from being updated, i.e. no relation = not updates from other components. Will this actually do the trick?
Is there a better way to ensuring there is no communication with DB while we are doing backup?

DnPlas replied:
Then we expect a downtime, perhaps we should mention that in the guide.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+   "source": [
+    "    # replace DB_CHARM with desired unit name: `katib-db` or `kfp-db`\n",
+    "    DB_CHARM= kfp-db | katib-db\n",
+    "    juju show-application kfp-db | yq '.[] | .charm'"
```
```suggestion
    "    juju show-application $DB_CHARM | yq '.[] | .charm'"
```

i-chvets replied:
Will update.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "# remove relation between katib-db-manager charm and katib-db charm\n",
+    "juju remove-relation katib-db-manager katib-db\n",
+    "\n",
+    "# connect to unit's IP address using mysql client and verify that `katib` database exists\n",
```
```suggestion
    "# connect to unit's IP address using mysql client and verify that `katib` database exists\n",
    This will log you into a mysql monitor promt...
```

i-chvets replied:
Will add.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "\n",
+    "# create backup of all databases file using `mysqldump` utility, username, password, and unit's IP address\n",
+    "# onbtained earlier\n",
+    "mysqldump --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD --column-statistics=0 --databases katib  > katib-db.sql\n",
```
Do we need to end the command with `;`?

```suggestion
    "mysqldump --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD --column-statistics=0 --databases katib  > katib-db.sql;\n",
```

i-chvets replied:
No need for `;`. This is utility that is executed on client machine that connects to MySQL, not SQL statement.

DnPlas replied:
okay

i-chvets commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,237 @@
+    "\n",
+    "# create backup of all databases file using `mysqldump` utility, username, password, and unit's IP address\n",
+    "# onbtained earlier\n",
+    "mysqldump --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD --column-statistics=0 --databases katib  > katib-db.sql\n",
```
Ok, I see the problem. After previous command `mysql . . . ` we need to exit. Will update the guide.
`mysqldump` is command line utility that is executed on the command line of client machine, not within `mysql` client session.

DnPlas replied:
Ok

i-chvets commented:
Updated with comments. Also added [markdown version](https://github.com/canonical/bundle-kubeflow/blob/doc-db-migration-guide/docs/db-migration-guide.md) for easy read.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+    "# this command will log you into mysql monitor prompt\n",
+    "mysql --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD\n",
+    "\n",
+    "# exit mysql client session\n",
```
I don't get it then, what is the purpose of this command? All this does (from what I can see) is log you in to the mysql monitor prompt. We seem to not get anything from it, if that's the case, let's ditch it.

DnPlas replied:
Oh, I think you added this to verify the integrity of the data.

i-chvets replied:
Same comment as before:
> Should provide SQL queries? How deep should we go into describing DB operations?
I'd would leave it outside of this guide, but open to suggestions.

DnPlas replied:
See my previous comment

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+    "DB_MYSQL_RELATION_ID=$(juju show-unit $DB_UNIT | yq '.[] | .relation-info | select(.[].endpoint == \"mysql\") | .[0] | .relation-id')\n",
+    "DB_USER=root\n",
+    "DB_PASSWORD=$(bash -c \"juju run --unit $DB_UNIT 'relation-get -r $DB_MYSQL_RELATION_ID - $DB_UNIT' | grep root_password\" | awk '{print $2}')\n",
+    "DB_IP=$(juju show-unit $DB_UNIT/0 | yq '.[] | .address')"
```
```suggestion
    "DB_IP=$(juju show-unit $DB_UNIT | yq '.[] | .address')"
```

And all where it applies

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+   "outputs": [],
+   "source": [
+    "# replace MYSQL_DB with desired application name: `katib-db-mysql` or `kfp-db-mysql`\n",
+    "MYSQL_DB= <katib-db-mysql/0 | kfp-db-mysql/0>\n",
+    "\n",
+    "# deploy new MySQL database charm\n",
+    "juju deploy mysql-k8s $MYSQL_DB --channel 8.0/stable --trust\n",
+    "\n",
+    "# obtain username and password of new MySQL database from MySQL charm\n",
+    "MYSQL_DB_USER=$(juju run-action $MYSQL_DB get-password --wait | yq '.[] | .results.username')\n",
+    "MYSQL_DB_PASSWORD=$(juju run-action $MYSQL_DB get-password --wait | yq '.[] | .results.password')\n",
+    "MYSQL_DB_IP=$(juju show-unit $MYSQL_DB | yq '.[] | .address')\n"
```
You have to be careful where exactly the unit number is used and when the application is used. For instance running `juju deploy mysql-k8s $MYSQL_DB --channel 8.0/stable --trust` after setting `MYSQL_DB=katib-db-mysql/0` will give you a name error.

i-chvets replied:
Updating to conver for this. Please see comment above about variables.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+    "\n",
+    "# create backup of all databases file using `mysqldump` utility, username, password, and unit's IP address\n",
+    "# onbtained earlier\n",
+    "mysqldump --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD --column-statistics=0 --databases katib  > katib-db.sql\n",
```
So the backup is a `.sql` file, worth mentioning that with this command a file will be generated and saved in the CWD.

i-chvets replied:
Will do.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+    "mysql --host=$MYSQL_DB_IP --user=$MYSQL_DB_USER --password=$MYSQL_DB_PASSWORD < katib-db.sql\n",
+    "\n",
+    "# relate katib-db-manager and new MySQL database charm\n",
+    "juju relate katib-db-manager:relational-db $MYSQL_DB:database"
```
Am I doing something wrong? I get this:

```
juju relate katib-db-manager:relational-db katib-db-mysql:database
ERROR application "katib-db-manager" has no "relational-db" relation
```

i-chvets replied:
Good point. Will add.

DnPlas commented on docs/db-migration-guide.ipynb:
```diff
@@ -0,0 +1,242 @@
+   "source": [
+    "# compare new MySQL database and compare it to the backup created earlier\n",
+    "mysqldump --host=$DB_IP --user=$DB_USER --password=$DB_PASSWORD --column-statistics=0 --databases katib  > katib-db-new.sql\n",
+    "diff katib-db.sql katib-db-new.sql\n",
```
What is the expected output of diff? Should it be empty?

DnPlas reviewed (commented):
Thanks @i-chvets, it's started to look better, I left some more comments, but I think we are getting closer to merge.

ColmBhandal reviewed (dismissed):
Aside from some technical details to be ironed out with @DnPlas, I'm happy with this. I reviewed on a call with Ivan and made some adjustments.

i-chvets commented:
There is a potential issue when primary key does not exist in the table.

i-chvets commented:
@DnPlas Ready for final review. When Data Platform team finishes their migration guide, we should link it with ours.
